cam6_2_018: CLUBB SILHS update

cacraigucar opened the pull request:
Closes #65 
Closes #77 

Update the CLUBB and SILHS libraries with updates from UWM

This is ready for the scientists to begin their climate runs and for the SEs to review.  

Note that currently the libraries are pointing to untagged commits for CLUBB and SILHS.  These will be tagged once UWM answers back with an issue I've had with SILHS (and I make the ChangeLogs for those repos).  

There may be a minor change for one to three SILHS namelist variables once UWM responds back.   I didn't want to hold up the testing and review of CLUBB while waiting for this fairly insignificant change for SILHS.

fvitt commented on bld/namelist_files/namelist_defaults_cam.xml:
```diff
@@ -22,8 +22,6 @@
 <dtime dyn="se"    hgrid="ne120np4"> 450</dtime>
 <dtime dyn="se"    hgrid="ne240np4"> 225</dtime>
 <dtime dyn="se"    hgrid="ne0np4CONUS.ne30x8">225</dtime>
```
It looks like this PR removes some of the refined grid settings in namelist_defaults.  Is this intended?

cacraigucar replied:
No.  That was an unintended change.  I'll work on fixing it.

gold2718 commented on src/physics/cam/clubb_intr.F90:
```diff
@@ -335,6 +351,11 @@ subroutine clubb_register_cam( )
+    call pbuf_add_field('THLP3',      'global', dtype_r8, (/pcols,pverp,dyn_time_lvls/), thlp3_idx)
+    call pbuf_add_field('UP3',        'global', dtype_r8, (/pcols,pverp,dyn_time_lvls/), up3_idx)
+    call pbuf_add_field('VP3',        'global', dtype_r8, (/pcols,pverp,dyn_time_lvls/), vp3_idx)
+
```
I hereby register my continued displeasure of the lazy use of PBUF variables for data which is only used in one module.

gold2718 commented on src/physics/cam/clubb_intr.F90:
```diff
@@ -96,27 +102,31 @@ module clubb_intr
+  real(r8) :: clubb_c_K9 = unset_r8
+  real(r8) :: clubb_nu9 = unset_r8
+  real(r8) :: clubb_c_K10 = unset_r8
+  real(r8) :: clubb_c_K10h = unset_r8
```
I know this is not new but is there any check in place for the case where any of these variables is _not_ in the namelist file? I do not think CLUBB has any internal checks.

cacraigucar replied:
Added checks

gold2718 commented on src/physics/cam/clubb_intr.F90:
```diff
@@ -950,49 +996,73 @@ subroutine clubb_ini_cam(pbuf2d)
+    endif
+
+    ! Print configurable CLUBB flags
+    write(iulog,'(a,i0,a)') " CLUBB configurable flags set in thread ", iam, ":"
```
It is bad form to spew CAM log messages into the CESM logfile. Is this write statement really necessary?

nusbaume replied:
I thought the `iulog` specification sent the text to the CAM (atm) log file, or is that only true when on masterproc?

gold2718 replied:
Right, it goes to the right file on masterproc but all the other tasks go to 6 which ends up (usually jumbled) in the CESM log file. There have been lots of complaints about components such as CAM polluting this file.

cacraigucar replied:
Removed write

gold2718 commented on src/physics/cam/clubb_intr.F90:
```diff
@@ -1866,11 +1941,11 @@ real(r8) :: exner_clubb(pcols,pverp)         ! Inverse exner function consistent
    !  surface pressure.  CAM's exner (in state does not).  Therefore, for consistent 
    !  treatment with CLUBB code, anytime exner is needed to treat CLUBB variables 
-   !  (such as thlm), use "exner_clubb" other wise use the exner in state
+   !  (such as thlm), use "inv_exner_clubb" other wise use the exner in state
```
otherwise is one word, I have no idea what 'other wise' is supposed to mean.

cacraigucar replied:
fixed

gold2718 commented on src/physics/cam/subcol_SILHS.F90:
```diff
@@ -272,21 +280,69 @@ subroutine subcol_init_SILHS(pbuf2d)
+                                                   silhs_config_flags )
+
+      ! Print the SILHS configurable flags
+      write(iulog,'(a,i0,a)') " SILHS configurable flags set in thread ", iam, ":"
```
Another broadcast to the CESM logfile? Is it necessary?

cacraigucar replied:
removed

cacraigucar commented on bld/namelist_files/namelist_defaults_cam.xml:
```diff
@@ -1277,31 +1278,35 @@
 <clubb_gamma_coef                                 > 0.308   </clubb_gamma_coef>
 <clubb_gamma_coef hgrid="1.9x2.5"                 > 0.280   </clubb_gamma_coef>
 <clubb_gamma_coefb                                > 0.32    </clubb_gamma_coefb>
+<clubb_beta                                       > 2.4     </clubb_beta>
```
I didn't  - UWM did.  These changes are all their mods.  I performed a little cleanup, but left everything mainly like they had it.

cacraigucar commented on src/physics/cam/clubb_intr.F90:
```diff
@@ -799,9 +833,21 @@ subroutine clubb_ini_cam(pbuf2d)
+
+    ! Allocate (in the vertical) and zero PDF parameters
+    do j = 1, pcols, 1
+       do l = begchunk, endchunk, 1
```
My gut feeling is that it won't make a difference since due to there being so much happening inside the subroutine calls.  Reordering couldn't hurt, so went ahead and  did it.

cacraigucar commented on src/physics/cam/clubb_intr.F90:
```diff
@@ -1866,11 +1941,11 @@ real(r8) :: exner_clubb(pcols,pverp)         ! Inverse exner function consistent
    !  Compute inverse exner function consistent with CLUBB's definition, which uses a constant
    !  surface pressure.  CAM's exner (in state does not).  Therefore, for consistent 
```
fixed

cacraigucar commented:
I believe I have addressed all of the review comments.  All testing is passing (with expected answer changes).  Awaiting final acceptance/further comments from reviewers and scientist's decision on climate runs.

cacraigucar commented on Externals_CAM.cfg:
```diff
@@ -23,14 +23,14 @@ required = True
 protocol = git
 repo_url = https://github.com/ESCOMP/SILHS_CESM
-tag = silhs_ncar_backwards_compat_20181205_c20191001
+tag = 555291e
```
If your review were just a little later, you wouldn't have even seen this.

I noted in my description that these were temporary placeholders (though they do work).  I am currently updating and tagging the CLUBB and SILHS repos.

Katetc replied:
Ha, ha, awesome, thanks!
